platforms/core-runtime/service-registry-impl/src/test/groovy/org/gradle/internal/service/DefaultServiceRegistryFromConstructorInjectionTest.groovy

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.service
+
+import spock.lang.Specification
+
+class DefaultServiceRegistryFromConstructorInjectionTest extends Specification {
+
+    TestRegistry registry = new TestRegistry()
+
+    def "can inject implementation via constructor"() {
+        given:
+        registry.addProvider(new ServiceRegistrationProvider() {
+            @Provides
+            TestService create(@FromConstructor TestServiceImpl service) { service }
```

Comment:
How about:

```java
@Provides
abstract TestService binding(TestServiceImpl service);
```

Reply:
This will require a more involved change that brings support for abstract Service Providers, which is currently not possible.

And there is a downside of not having the compiler check the `instanceof` link between the implementation and the service interface, which happens now due to having the dummy implementation of just returning the implementation instance.

On the other hand, the team shared a sentiment to get rid of the naming conventions altogether, and just rely on the presence of the `@Provides` annotation on the method. In this sense, all the provider methods will be "equal", and any behavior changes should be explicitly encoded in the parameters themselves.

I support that sentiment, and I would rather move in that direction.

Another example of this is to abandon the `decorate` naming pattern, and instead have something like a `@Decorated` / `@FromParent` annotation on the corresponding argument to make it very explicit.